File: pkg-py/tests/test_remote_wikipathways.py
"""WikiPathways remote mode: live existence checks via the published asset."""

import pytest

import biobouncer
import biobouncer._remote as remote
from biobouncer import RemoteError


@pytest.fixture(autouse=True)
def _isolate_cache(tmp_path, monkeypatch):
    monkeypatch.setenv("BIOBOUNCER_CACHE_DIR", str(tmp_path))


def _stub(status, body=None):
    def _http_get(url, timeout=30):
        return status, body

    return _http_get


def test_url_builds_the_asset_endpoint():
    from biobouncer._registry import get_source

    url = remote._wikipathways_url(get_source("wikipathways"), "WP554")
    assert url == (
        "https://www.wikipathways.org/wikipathways-assets/pathways/WP554/WP554.gpml"
    )


def test_existing_pathway_is_valid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biobouncer.check_id("WP554", source_db="wikipathways", how="remote")[0]
    assert res.valid is True
    assert res.normalized == "WP554"
    assert res.suggestion is None


def test_absent_pathway_is_invalid(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(404))
    res = biobouncer.check_id("WP9999999", source_db="wikipathways", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_lowercase_suggests_the_uppercase_form(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(200))
    res = biobouncer.check_id("wp554", source_db="wikipathways", how="remote")[0]
    assert res.valid is False
    assert res.suggestion == "WP554"


def test_malformed_skips_the_network(monkeypatch):
    def _forbidden(url, timeout=30):
        raise AssertionError("a malformed id must not reach the network")

    monkeypatch.setattr(remote, "_http_get", _forbidden)
    res = biobouncer.check_id("WPXYZ", source_db="wikipathways", how="remote")[0]
    assert res.valid is False
    assert res.suggestion is None


def test_unexpected_status_raises_and_is_not_cached(monkeypatch):
    monkeypatch.setattr(remote, "_http_get", _stub(500))
    with pytest.raises(RemoteError):
        biobouncer.check_id("WP554", source_db="wikipathways", how="remote")
    path = remote._remote_cache_path("wikipathways", "pathways", "WP554")
    assert not path.is_file()

File: pkg-r/tests/testthat/test-remote-wikipathways.R
# Offline tests for wikipathways remote mode. Existence is checked against the
# published WikiPathways asset; the biogate.remote_transport option replaces it.

.stub_wikipathways <- function(status, body = NULL) {
  function(url, timeout) list(status = status, body = body)
}

test_that("the url addresses the asset endpoint", {
  url <- .remote_resolvers$wikipathways$url(NULL, "WP554")
  expect_identical(
    url,
    "https://www.wikipathways.org/wikipathways-assets/pathways/WP554/WP554.gpml"
  )
})

test_that("an existing pathway is valid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_wikipathways(200))
  res <- check_id("WP554", "wikipathways", how = "remote")
  expect_true(res$valid)
  expect_identical(res$normalized, "WP554")
  expect_true(is.na(res$suggestion))
})

test_that("an absent pathway is invalid", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_wikipathways(404))
  res <- check_id("WP9999999", "wikipathways", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("a lowercase id suggests the uppercase form", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_wikipathways(200))
  res <- check_id("wp554", "wikipathways", how = "remote")
  expect_false(res$valid)
  expect_identical(res$suggestion, "WP554")
})

test_that("a malformed id skips the network", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(
    biogate.remote_transport = function(url, timeout) {
      stop("a malformed id must not reach the network")
    }
  )
  res <- check_id("WPXYZ", "wikipathways", how = "remote")
  expect_false(res$valid)
  expect_true(is.na(res$suggestion))
})

test_that("an unexpected status raises and is not cached", {
  withr::local_envvar(BIOGATE_CACHE_DIR = withr::local_tempdir())
  withr::local_options(biogate.remote_transport = .stub_wikipathways(500))
  expect_error(
    check_id("WP554", "wikipathways", how = "remote"),
    class = "biogate_error_remote"
  )
  expect_false(
    file.exists(.remote_cache_path("wikipathways", "pathways", "WP554"))
  )
})
